File: ungrouped/3112.py
import heapq

# TLE
class Solution:
    def minimumTime(self, n: int, edges: List[List[int]], disappear: List[int]) -> List[int]:

        # Create an adjacency list to represent the graph
        graph = {i: [] for i in range(n)}
        for u, v, length in edges:
            graph[u].append((v, length))
            graph[v].append((u, length))
        
        # Initialize distances array with -1 (unreachable)
        distances = [-1] * n
        distances[0] = 0  # Distance from node 0 to itself is 0
        
        # Priority queue to store nodes to visit
        pq = [(0, 0)]  # (distance, node)
        
        while pq:
            dist, node = heapq.heappop(pq)
            
            # If the node has disappeared, skip it
            if disappear[node] <= dist:
                continue
            
            for neighbor, length in graph[node]:
                # If the neighbor has disappeared, continue
                if disappear[neighbor] <= dist + length:
                    continue
                
                # If the neighbor is reachable or has never been visited
                if distances[neighbor] == -1 or distances[neighbor] > dist + length:
                    distances[neighbor] = dist + length
                    heapq.heappush(pq, (dist + length, neighbor))
        
        return distances


class Solution:
    def minimumTime(self, n: int, edges: List[List[int]], disappear: List[int]) -> List[int]:
        # Create an adjacency list to represent the graph
        graph = {i: [] for i in range(n)}
        for u, v, length in edges:
            graph[u].append((v, length))
            graph[v].append((u, length))

        # Initialize distances array with -1 (unreachable)
        distances = [-1] * n
        distances[0] = 0  # Distance from node 0 to itself is 0

        # Priority queue to store nodes to visit
        pq = [(0, 0)]  # (distance, node)

        # Set to store visited nodes
        visited = set()

        while pq:
            dist, node = heapq.heappop(pq)

            # If the node has disappeared or is already visited, skip it
            if disappear[node] <= dist or node in visited:
                continue

            visited.add(node)

            for neighbor, length in graph[node]:
                # If the neighbor has disappeared or is already visited, continue
                if disappear[neighbor] <= dist + length or neighbor in visited:
                    continue

                # If the neighbor is reachable or has never been visited
                if distances[neighbor] == -1 or distances[neighbor] > dist + length:
                    distances[neighbor] = dist + length
                    heapq.heappush(pq, (dist + length, neighbor))

        return distances

File: ungrouped/3112.cpp
#include "include.h"

// TLE
//Dijkstra’s Algorithm
class Solution {
public:
    vector<int> minimumTime(int n, vector<vector<int>>& edges, vector<int>& disappear) {
        vector<vector<pair<int, int>>> graph(n, vector<pair<int, int>>());
        for(int i = 0; i < edges.size(); i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            int w = edges[i][2];
            graph[u].push_back({v, w});
            graph[v].push_back({u, w});
        }

        vector<int> res(n, -1);
        // Start from 0
        res[0] = 0;

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        pq.push({0, 0});
        while(!pq.empty()) {
            auto [dist, node] = pq.top();
            pq.pop();
            
            if(disappear[node] <= dist) {
                continue;
            }

            for(auto [next, weight] : graph[node]) {
                if(disappear[next] <= dist + weight) {
                    continue;
                }

                if(res[next] == -1 || res[next] > dist + weight) {
                    res[next] = dist + weight;
                    pq.push({dist + weight, next});
                }
            }
        }
        return res;
    }
};

// Dijkstra’s Algorithm + memory
class Solution {
public:
    vector<int> minimumTime(int n, vector<vector<int>>& edges, vector<int>& disappear) {
        vector<vector<pair<int, int>>> graph(n, vector<pair<int, int>>());
        for(int i = 0; i < edges.size(); i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            int w = edges[i][2];
            graph[u].push_back({v, w});
            graph[v].push_back({u, w});
        }

        vector<int> res(n, -1);
        // Start from 0
        res[0] = 0;
        vector<int> visited(n, 0);

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        pq.push({0, 0});
        while(!pq.empty()) {
            auto [dist, node] = pq.top();
            pq.pop();
            
            // disappear or visited
            if(disappear[node] <= dist || visited[node]) {
                continue;
            }

            visited[node] = 1;

            for(auto [next, weight] : graph[node]) {
                if(disappear[next] <= dist + weight || visited[next]) {
                    continue;
                }
                
                if(res[next] == -1 || res[next] > dist + weight) {
                    // update neighbors
                    res[next] = dist + weight;
                    pq.push({dist + weight, next});
                }
            }
        }
        return res;
    }
};